internal/stage_hp3.go

```diff
@@ -0,0 +1,181 @@
+		return err
+	}
+	defer os.Remove(historyFile)
+
+	utils.LogReadableFileContents(logger, initialContent, "Original history file content:", filepath.Base(historyFile))
+
+	// Set HISTFILE to /dev/null before starting the shell
+	shell.Setenv("HISTFILE", "/dev/null")
+
+	if err := asserter.StartShellAndAssertPrompt(true); err != nil {
+		return err
+	}
+
+	// Step 2: Run some commands in the shell
+	nShellCommands := random.RandomInt(2, 4)
+	commandTestCases := make([]test_cases.CommandResponseTestCase, nShellCommands)
+	for i := 0; i < nShellCommands; i++ {
+		cmdWords := random.RandomWords(random.RandomInt(2, 4))
+		cmd := "echo " + strings.Join(cmdWords, " ")
+		commandTestCases[i] = test_cases.CommandResponseTestCase{
+			Command:        cmd,
+			ExpectedOutput: strings.Join(cmdWords, " "),
+			SuccessMessage: fmt.Sprintf("✓ Ran %s", cmd),
+		}
+	}
+
+	// Step 3: Check history before appending
+	historyBeforeTest := test_cases.HistoryTestCase{
+		CommandsBeforeHistory: commandTestCases,
+		SuccessMessage:        "✓ History before appending is correct",
+	}
+	if err := historyBeforeTest.Run(asserter, shell, logger); err != nil {
+		return err
+	}
+
+	// Step 4: Append history to file
+	historyAppendCmd := "history -a " + historyFile
+	historyAppendTest := test_cases.CommandReflectionTestCase{
+		Command:        historyAppendCmd,
+		SuccessMessage: "✓ History appended to file",
+	}
+	if err := historyAppendTest.Run(asserter, shell, logger, false); err != nil {
+		return err
+	}
+
+	// Read the updated history file content
+	historyContent, err := os.ReadFile(historyFile)
+	if err != nil {
+		logger.Errorf("Failed to read history file: %v", err)
+		return err
+	}
+
+	utils.LogReadableFileContents(logger, string(historyContent), "History file content after appending:", filepath.Base(historyFile))
+
+	// Check if all commands are present in the history file
+	historyStr := strings.TrimSpace(string(historyContent))
+	historyLines := strings.Split(historyStr, "\n")
+	if len(historyLines) != len(initialCommands)+len(commandTestCases)+2 {
+		return fmt.Errorf("history file has %d lines, expected %d", len(historyLines), len(initialCommands)+len(commandTestCases)+2)
+	}
+	logger.Successf("✓ History file has correct number of lines")
+	i := 0
+	// First verify initial commands are still present
+	for _, cmd := range initialCommands {
+		if historyLines[i] != cmd {
+			return fmt.Errorf("expected initial command %q at line %d, got %q", cmd, i+1, historyLines[i])
+		}
+		logger.Successf("✓ Found initial command %q in history file", cmd)
+		i++
+	}
+
+	// Then verify new commands were appended
+	for _, cmd := range commandTestCases {
+		if historyLines[i] != cmd.Command {
+			return fmt.Errorf("expected new command %q at line %d, got %q", cmd.Command, i+1, historyLines[i])
+		}
+		logger.Successf("✓ Found new command %q in history file", cmd.Command)
+		i++
+	}
+
+	// Verify history command itself is present
+	if historyLines[i] != "history" {
```

Comment:
Really think you should extract all of this historyFile stuff into a re-usable thing - there's a LOT of duplication here